frontend/lib/resolution/scope-queries.cpp

```diff
@@ -2740,16 +2772,115 @@ struct GatherUsedImportedIds {
   void exit(const AstNode* ast) { }
 };
 
-static const std::vector<ID>&
-findAllModulesUsedImportedInTreeQuery(Context* context, ID modId) {
-  QUERY_BEGIN(findAllModulesUsedImportedInTreeQuery, context, modId);
+// save the module used/imported to idSet / idVec
+void GatherMentionedModules::gatherModuleId(const ID& id) {
+  // filter out enums, other stuff, since they are not of interest here
+  if (parsing::idIsModule(context, id)) {
+    auto p = idSet.insert(id);
+    if (p.second) {
+      // insertion occured, so add it also to the vector
+      idVec.push_back(id);
+    }
+  }
+}
+
+// Gather any modules in vec
+// Return the first Scope* for a module in vec
+// Ambiguity should be reported elsewhere so is ignored here.
+const Scope*
+GatherMentionedModules::gatherIdsAndFindScope(const VecBorrowedIds& v) {
+  const Scope* scope = nullptr;
+  for (const auto& bids : v) {
+    for (const auto& id : bids) {
+      if (scope == nullptr && parsing::idIsModule(context, id)) {
+        scope = scopeForModule(context, id);
+      }
+      gatherModuleId(id);
+    }
+  }
+
+  return scope;
+}
+
+// gather any modules mentioned in a Dot sequence or Identifier
+// and also return the Scope for that Dot sequence or Identifier.
+const Scope*
+GatherMentionedModules::gatherAndFindScope(const Scope* scope,
+                                           const AstNode* ast) {
+  const LookupConfig config = LOOKUP_DECLS |
+                              LOOKUP_IMPORT_AND_USE |
+                              LOOKUP_PARENTS |
+                              LOOKUP_INNERMOST;
+
+  if (auto ident = ast->toIdentifier()) {
+
+    // handle super/this just in case this processes import Dot exprs
+    if (ident->name() == USTR("super")) {
+      return nullptr;
+    } else if (ident->name() == USTR("this")) {
+      return scope->moduleScope();
+    }
+
+    auto r = lookupNameInScope(context, scope, /* receiver scopes */ {},
+                               ident->name(), config);
```

Comment:
Not setting `receiver scopes` might be a problem:

```Chapel
module M {

}
module N {
  record MatrixContainer {
    var M: Matrix(...);
   
    proc foo() do return M;
  }
}
```

Reply:
This function `gatherAndFindScope` is only called when processing a `Dot` expression and your example does not have a `Dot`. Additionally, the way the lookup is configured, it won't go up beyond `module N` so it wouldn't find `module M`. It would just find nothing, which is not a problem here.

However, here is a program that does reveal the issue:

``` chapel
      module Program {
        module M {
          var M: real;
        }
        
        record r1 { var M: int; }
        record r2 { var M: r1; }

        proc r2.foo() {
          return M.M;
        }

        proc main() {
          var x: r2;
          var y = x.foo();
        }
      }
```

I'll think about this some more. On the one hand, using the scope resolver here would allow it to be accurate. On the other, I'm worried that using the scope resolver here will cause performance issues and prevent progress on the PR that this PR is preparing for.

Reply:
I have something that uses the scope resolver in such method cases to be more correct here. It adds 0.02s to a the `testModuleInitOrder` part computing the module init order for the bundled modules. This seems acceptable for now.